### PR TITLE
`MultiAddressClientBuilder` execution strategy control

### PR DESCRIPTION
**Pull request by bondolo**
Motivation:
The `MultiAddressClientBuilder` does not currently always respect
overrides of execution strategy made by the single address
initializer
Modification:
The computed execution strategy is a combination of the multiaddress
client builder strategy, the api strategy and the client strategy.
Result:
More predictable and controllable behavior for `MultiAddressClientBuilder`

### CODE REVIEW

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:288]: We have to implement `closeGracefully()` method too

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:272]: Instead of wrapping manually, consider using a filter that you can append before invoking `singleAddressInitializer`. That way you won't need to delegate/implement most of the methods, it should be enough to implement only `request` method.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:409]: nested ternary operators are not readable, please use `if` with brackets

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:414]: Last time we discussed this problem, we agreed to keep it simple: take whatever `HTTP_EXECUTION_STRATEGY_KEY` comes from multi-address level and merge it with a `singleClientStrategy` to make sure we do not miss a case when single-address client requires more offloading. We agreed to avoid an extra key, like `HTTP_CLIENT_API_KEY`, and any dependency on API variant. Can you clarify why this was necessary to consider API at the end?

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpContextKeys.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpContextKeys.java:39]: The `HttpContextKeys` is a holder for all public keys we expect users to use. If we ever add a special key for our internal purposes, we should not make it public.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpApiConversions.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpApiConversions.java:35]: I'm not in favor of expanding public API. We have a good opportunity to get rid of `HttpApiConversions` instead of expanding it.

**Review by idelpivnitskiy — DISMISSED** [review]

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java
Comment (bondolo) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:191]: I do agree that it is confusing. There should at least be a warning that the filter chain specifies offloads whereas the builder strategy does not.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (bondolo) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:271]: I don't understand how that would work.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:297]: The default is defined in `GracefulAutoCloseable` and invokes `close()`. It will result in non-graceful closure if the method is not implemented and delegated.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:271]: Implement a regular `StreamingHttpClientFilterFactory` that returns a `StreamingHttpClientFilter` as we usually do and append it before invoking an initializer.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java
Comment (idelpivnitskiy) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:191]: #2189

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (bondolo) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:297]: `StreamingHttpClient` also provides a default which does a graceful close. This may be moot if the filter approach is used though.

Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:297]: Sorry for confusion, when I posted the original comment, I had `AsyncCloseable#closeAsyncGracefully()` in mind instead of `GracefulAutoCloseable#closeGracefully()`

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:276]: when we implement a filter, the recommended approach is to use `delegate.request(request)` to avoid any unexpected behavior changes if the `super` implementation changes

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:275]: 1. I believe we should wrap it with `defer` to make sure we check the strategy on each re-subscribe.
2. You can take `clientStrategy` from `client.executionContext().executionStrategy()`, no need to have a class-level variable.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:279]: Consider moving it to `MultiAddressStrategyWrapper` to make it static

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:289]: Intellij IDEA suspects NPE here. Consider adding `assert requestStrategy != null;` after you take `requestStrategy` from the `contextMap`

**Review by idelpivnitskiy — COMMENTED** [review]
LGTM after comments addressed:

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (bondolo) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:279]: Since it is only used once I removed the argument and use the field.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:285]: `singleStrategy` should be taken from `client` (line 274)

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:271]: After you remove this variable, consider making this class a singleton

**Comment by idelpivnitskiy** [comment]
Will have a final look after merge conflicts resolved

**Comment by bondolo** [comment]
Temporarily draft as pieces are pulled into separate standalone PRs.

**Comment by idelpivnitskiy** [comment]
After #2192 I saw a new flaky case:
- https://github.com/apple/servicetalk/pull/2247/checks?check_run_id=6804389274
- https://github.com/apple/servicetalk/pull/2243/checks?check_run_id=6799380558
- https://github.com/apple/servicetalk/pull/2243/checks?check_run_id=6799141984

**Comment by bondolo** [comment]
> After #2192 I saw a new flaky case:
> 
> * https://github.com/apple/servicetalk/pull/2247/checks?check_run_id=6804389274
> * https://github.com/apple/servicetalk/pull/2243/checks?check_run_id=6799380558
> * https://github.com/apple/servicetalk/pull/2243/checks?check_run_id=6799141984

I have switched the test to use `ExecutionMode.SAME_THREAD` to attempt to isolate this.

**Review by idelpivnitskiy — APPROVED** [review]
🚀

**Comment by idelpivnitskiy** [comment]
Hey, some outdated comments from intermediate iterations were published unintentionally. All of those are already resolved. Go ahead with merging the PR. Thanks for the hard work on it!

**Pull request merged**
